doc/dev/error-categorization.md

```diff
@@ -0,0 +1,173 @@
+error/warning in there.
+
+For a complete structural overview please visit the corresponding [design
+decision document](../decisions/error_codes.md)
+
+### Permanent errors ("C01000")
+
+The branch category `Permanent Errors` refer to such errors which cannot be
+fixed by retry at all. `Permanent Errors` are subdivided into
+
+- Resource
+- Installation
+- Logical
+
+#### Resource ("C01100")
+
+`Resource Errors` as a branch category are all kinds of errors which are either
+permission related, existence related or missing resources such as memory, RAM,
+etc. Examples are missing files, insufficient permissions or out of memory. This
+category forces users or applications to provide additional resources, create a
+missing file/ directory etc or to provide the relevant permissions.
+
+##### Memory Allocation ("C01110")
+
+`Memory Allocation Errors` are errors which come from failed `elektraMalloc`
+calls primarily as no more memory could be allocated for the application. Errors
+with not enough hard disc space do not belong here but into `General Resource`.
+Such errors will gain special handling in future releases and users cannot deal
+with such errors as of now.
+
+##### General Resource ("C01120")
+
+`General Resource Errors` are all kind of permission, existence and resource
+errors which are essential for Elektra to operate. Compared to validation errors
+this category has its focus the underlying system resources whereas validation
+errors for usages with specifications. Examples are missing files/ directories
+or insufficient permission to execute certain commands (eg. you would require
+sudo permissions). Reactions are fixing the permissions, remount, creating the
+file/directory and retry the operation. Compared to validation errors,
+administrators would change the specification or retry with a different value.
+
+#### Installation ("C01200")
+
+`Installation Errors` are those errors which are related to a wrong installation
+such as wrong plugin names, missing backends, initialization errors,
+misconfiguration of Elektra etc. Installation errors might also be non-Elektra
+specific but also from dependent library/applications such as gpg. Also plugin
+configuration errors belong to `Installation Errors` as this happens during
+mounting. Users will have to reconfigure, reinstall, recompile (with other
+settings) Elektra in order to get rid of this error or fix the installation of
+the corresponding library/application.
+
+#### Logical ("C01300")
+
+`Logical Errors` is a branch category in which you indicate a bug in Elektra
+such as internal errors, assertion failures, not implemented features or errors
+which should not happen such as going into a `default` branch when you are
+assured that all cases are covered. Usually such errors come with a message to
+report such failures to Elektra's bugtracker. Applications cannot handle such
+errors themselves.
+
+##### Internal ("C01310")
```

Comment:
Will be merged to `Logical` since branches are now allowed now

Reply:
See above. And we need a distinction between errors that are bugs in core, in plugins and API use.

Reply:
Why do we suddenly need the distinction? The reaction does not differ.

The categorization was more or less clear and stable as I changed it very often due to the previous PR and I was only expecting minor changes but suddenly you want to completely restructure everything again? The `assertion` errors by the way were initially suggested by you and made sense but suddenly you want them to be removed? This is getting very confusing as this proceeds. Please let us not reinvent the wheel twice. The categorization will not be perfect, I honestly expect it to take several month to years to be near to perfect as we also do not have notable experience on what users of elektra expect of errors (or developers) and it will change in future versions no matter how often I rejig this.

Reply:
> Maybe we find a better name than "Internal". Errors by design are never internal (if we can handle them internally, we would never create an error. See above for suggestion.).

I honestly find the name internal absolutely correct because `Assertion` or `Internal` errors are such errors which you simply do not expect but have to deal with. As written in 2 other posts: default branches which should not be taken are such examples. You cannot "handle" these errors internally because you are sure they do not happen and if they do you still cannot say why and what to do.

Also the programmatic reaction is clear because it clearly indicates a bug where the source was not the user but a misimplemented feature.

Reply:
> Why do we suddenly need the distinction? The reaction does not differ.

I would not suggest it if I saw no difference. For bugs in plugins, the user can simply mount another plugin which behaves better, for bugs in the core this is not possible.

But I agree it is a subtle difference (in particular: what is about default cases in plugins).

> The categorization was more or less clear and stable as I changed it very often due to the previous PR and I was only expecting minor changes but suddenly you want to completely restructure everything again?

I do not understand where you see a "complete restructure" if I only see the need of clarification of two categories (resource and logic).

> also do not have notable experience on what users of elektra expect of errors (or developers)

That is why this process takes so long. It is *essential* that you get expert on your topic.

> and it will change in future versions no matter how often I rejig this

Of course there might be some need to extend it later. But as we saw in PostgreSQL: in well designed error categories this is possible without breaking anything.

> [the name internal errors]

Ok, then internal it is.